File: server/services/youtube-feed-service/src/main/java/com/richo/reader/youtube_feed_service/FeedCache.java
package com.richo.reader.youtube_feed_service;

import com.richodemus.reader.dto.FeedId;
import com.richodemus.reader.dto.FeedName;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.inject.Inject;
import javax.inject.Singleton;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;
import java.util.Optional;
import java.util.Set;

@Singleton
public class FeedCache {
    private final Logger logger = LoggerFactory.getLogger(getClass());
    private final Map<FeedId, Feed> cache;
    private final JsonFileSystemPersistence fileSystemPersistence;

    @Inject
    public FeedCache(JsonFileSystemPersistence fileSystemPersistence) {
        this.fileSystemPersistence = fileSystemPersistence;
        this.cache = new HashMap<>();
    }

    public Optional<Feed> get(FeedId feedId) {
        logger.debug("Fetching channel {}", feedId);
        return Optional.ofNullable(cache.computeIfAbsent(feedId, name -> fileSystemPersistence.getChannel(feedId).orElse(null)));
    }

    public void update(Feed feed) {
        logger.debug("Updating feed: {}, {} items", feed.getId(), feed.getItems().size());
        cache.put(feed.getId(), feed);
        fileSystemPersistence.updateChannel(feed);
    }

    Set<FeedId> getAllFeedIds() {
        return cache.keySet();
    }

    void add(FeedId feedId) {
        logger.info("Adding feed {}", feedId);
        cache.putIfAbsent(feedId, new Feed(feedId, new FeedName("UNKNOWN_FEED"), new ArrayList<>(), 0L));
    }
}

File: server/services/youtube-feed-service/src/main/kotlin/com/richo/reader/youtube_feed_service/FeedCache.kt
package com.richo.reader.youtube_feed_service

import com.richodemus.reader.dto.FeedId
import com.richodemus.reader.dto.FeedName
import org.slf4j.LoggerFactory
import java.util.ArrayList
import java.util.Optional
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
internal class FeedCache
@Inject constructor(private val fileSystemPersistence: JsonFileSystemPersistence) {
    private val logger = LoggerFactory.getLogger(javaClass)
    private val cache = mutableMapOf<FeedId, Feed>()

    fun getAllFeedIds() = cache.keys

    operator fun get(feedId: FeedId): Optional<Feed> {
        logger.debug("Fetching channel {}", feedId)
        return Optional.ofNullable(cache.computeIfAbsent(feedId) {
            fileSystemPersistence.getChannel(feedId).orElse(null)
        })
    }

    fun update(feed: Feed) {
        logger.debug("Updating feed: {}, {} items", feed.id, feed.items.size)
        cache.put(feed.id, feed)
        fileSystemPersistence.updateChannel(feed)
    }

    fun add(feedId: FeedId) {
        logger.info("Adding feed {}", feedId)
        cache.putIfAbsent(feedId, Feed(feedId, FeedName("UNKNOWN_FEED"), ArrayList(), 0L))
    }
}
